src/multigpu/mod.rs

```diff
@@ -0,0 +1,206 @@
+            // not fully initialized.
+            .filter(|dir| {
+                fs::File::create(dir.path())
+                    .map(|f| f.try_lock_exclusive().is_err())
+                    .unwrap_or(false)
+            })
+            // Deserialize file names to Tasks
+            .filter_map(|dir| {
+                dir.file_name()
+                    .into_string()
+                    .ok()
+                    .and_then(|name| Task::from_str(&name).ok())
+            })
+            .collect::<Vec<_>>())
+    }
+
+    // Schedule `f` on one of the `self.resources`
+    pub fn schedule<F, T>(&mut self, task: Task, f: F) -> SchedulerResult<T>
+    where
+        F: FnOnce(&R) -> T,
+    {
+        // We will keep all of our locked task files here, so that whenever `schedule` function
+        // is exited (Peacefully, or because of an error/crash), all of its associated Task files
+        // (And thus their locks) are dropped automatically and they no longer will be considered
+        // as active tasks and won't show up when `tasks_of()` is called.
+        let mut lock_files = HashMap::<PathBuf, fs::File>::new();
+
+        loop {
+            // Check if we can find a free device before enqueueing task files.
+            for resource in self.resources.iter() {
+                let tasks = self.tasks_of(resource)?;
+                if task.has_priority_over(&tasks) {
+                    // Although the given task has priority over all of the tasks currently queued
+                    // on the resource, we can't use it until the corresponding lock file of that
+                    // device is free. (I.e we can have the priority, but it doesn't mean that the
+                    // lower-priority Task has released the resource yet)
+                    // Having a separate lock-file for the resource, gives us the guarantee that
+                    // no two tasks can use the resource at the same time.
+                    let resource_lock_path = self.root.join(resource.id()).join("LOCK");
+                    match fs::File::create(resource_lock_path)?.try_lock_exclusive() {
+                        Ok(lock) => {
+                            // This is not needed, and it's just for the purpose of keeping
+                            // our scheduler directory clean.
+                            for (path, _) in lock_files.iter() {
+                                fs::remove_file(path)?;
+                            }
+
+                            let result = f(resource);
+                            drop(lock);
+                            return Ok(result);
+                        }
+                        Err(_) => {
+                            continue;
+                        }
+                    }
+                }
+            }
+
+            // No free devices found! Enqueue the task into all of the devices.
+            for resource in self.resources.iter() {
```

Comment:
Every task should not be scheduled on every resource. Resources should be specified per task.